### ISSUE TITLE
Looking for Deep sort or any good tracker which supports yolov4x-mish,csp/swish trained model weight.

### ISSUE THREAD

**Issue by akashAD98** [AlexeyAB/darknet#8064]
looking for any script/ support  (python,opencv,darknet etc)for this trained model.I have trained object detection using this model, Now I want to integrate a tracker on the top of this detection. If anyone has tried this please let us know. Thanks

**Comment by minhtcai** [comment]
I've been using SORT for my personal project,  this is an extension of darknet_video.py hope this could help. Will make a pull request soon.
https://github.com/minhtcai/darknet

**Comment by haviduck** [comment]
> I've been using SORT for my personal project, this is an extension of darknet_video.py hope this could help. Will make a pull request soon.
> https://github.com/minhtcai/darknet

hey i like your implementation thanks for sharing. i patched sort to accept class ids, so you dont have to fire up multiples.  

pretty just pass the cid to the update and it returns. works great but not optimized.

https://gist.github.com/haviduck/d7037eb7eb8d8d55cbfe6b3c2b522fc6

```
      newlist.append([xmin, ymin, xmax, ymax, c, dnum])
      npa = np.array(newlist)
trackers = self.mot_tracker.update(npa)
for d in trackers:
            d = d.astype(np.int32)
            classid= d[5]
```

**Comment by akashAD98** [comment]
@minhtcai thanks for your work, & here its google collab for the same.
https://colab.research.google.com/drive/1wX3fk_GiOpHSnPkxXHKCU7bEcr5v8UO0?usp=sharing

**Comment by minhtcai** [comment]
> > I've been using SORT for my personal project, this is an extension of darknet_video.py hope this could help. Will make a pull request soon.
> > https://github.com/minhtcai/darknet
> 
> hey i like your implementation thanks for sharing. i patched sort to accept class ids, so you dont have to fire up multiples.
> 
> pretty just pass the cid to the update and it returns. works great but not optimized.
> 
> https://gist.github.com/haviduck/d7037eb7eb8d8d55cbfe6b3c2b522fc6
> 
> ```
>       newlist.append([xmin, ymin, xmax, ymax, c, dnum])
>       npa = np.array(newlist)
> trackers = self.mot_tracker.update(npa)
> for d in trackers:
>             d = d.astype(np.int32)
>             classid= d[5]
> ```

Amazing @haviduck thank you very much!
I was trying to add your work inside this file https://github.com/minhtcai/darknet/blob/master/darknet_tracker_havi.py

Not sure I fully understand it, so for line 108, I should add a step to map string class to class ids, am I correct?
Could you please take a look or make a suggestion/pull request if possible? :D

**Comment by minhtcai** [comment]
> @minhtcai thanks for your work, can you add deep_sort for the same?
> I think deep sort has good accuracy compared to sort & here its google collab for the same.
> 
> https://colab.research.google.com/drive/1wX3fk_GiOpHSnPkxXHKCU7bEcr5v8UO0?usp=sharing

I'm trying to add deep sort in, I think it could take 1-2 weeks :D

**Comment by haviduck** [comment]
> > > I've been using SORT for my personal project, this is an extension of darknet_video.py hope this could help. Will make a pull request soon.
> > > https://github.com/minhtcai/darknet
> > 
> > 
> > hey i like your implementation thanks for sharing. i patched sort to accept class ids, so you dont have to fire up multiples.
> > pretty just pass the cid to the update and it returns. works great but not optimized.
> > https://gist.github.com/haviduck/d7037eb7eb8d8d55cbfe6b3c2b522fc6
> > ```
> >       newlist.append([xmin, ymin, xmax, ymax, c, dnum])
> >       npa = np.array(newlist)
> > trackers = self.mot_tracker.update(npa)
> > for d in trackers:
> >             d = d.astype(np.int32)
> >             classid= d[5]
> > ```
> 
> Amazing @haviduck thank you very much!
> I was trying to add your work inside this file https://github.com/minhtcai/darknet/blob/master/darknet_tracker_havi.py
> 
> Not sure I fully understand it, so for line 108, I should add a step to map string class to class ids, am I correct?
> Could you please take a look or make a suggestion/pull request if possible? :D

yeah man, but its been a while since i wrote this and its 4 am on a friday and im seeing double. so i think this should be all you need, but if not ill just send you everything.


```
namesPath = "./data/obj.names"

def read_obj_names(name_path):
    # read .names file, return dict mapping name to int
    # first line is number 1, second is 2, etc
    name_to_num = dict()
    obj_num = 1
    with open(name_path) as f:
        for line in f:
            obj_name = str(
                line
            ).strip()  # remove newline (and other leading/trailing whitespace)
            name_to_num[obj_name] = obj_num
            obj_num += 1
    return name_to_num

def number_to_letter(n):
    return number_to_letter_mapping[n]


def det_type_to_number(det_type):
    """map det_type to a numeric identifier, or a default (42) if det_type is missing in mapping dict"""
    if det_type in det_type_mapping:
        return det_type_mapping[det_type]["number"]


def det_type_number_to_color(det_type_number):
    """map a det_type_number to a color, defaulting to [grey] if the det_type_number is not found"""
    if det_type_number in det_type_number_to_color_mapping:
        return det_type_number_to_color_mapping[det_type_number]
    else:
        return [42, 42, 42]

det_type_mapping = {
    "Car": {"number": 1, "color": [0, 255, 0]},
    "Stephen Hawking": {"number": 2, "color": [255, 0, 0]},
    "Trumps Hair": {"number": 3, "color": "super weird"},
}

det_type_to_number_mapping = read_obj_names(namesPath)
if len(det_type_to_number_mapping) == 0:
    print("WARNING: no names found in {names_path}".format(names_path=namesPath))
# turn det_type_to_number_mapping inside out, mapping number to det_type
number_to_det_type_mapping = {
    number: det_type for det_type, number in det_type_to_number_mapping.items()
}
########### / map detection type ############
#
# flip det_type_mapping to a mapping from det_type_number to color for text rendering
det_type_number_to_color_mapping = {
    det_type_spec["number"]: det_type_spec["color"]
    for det_type_spec in det_type_mapping.values()
}

```

and then in your for loop
```
for blabla in tracker:

       det_type_number = blabla[5]
       nameofthethingamaging = number_to_det_type_mapping[det_type_number]
       color = det_type_number_to_color(det_type_number)
```
 Ive added a tripline counter to this aswell if that is of interest.

**Comment by haviduck** [comment]
follow up, after this i looked into the code and there is a clear idswitch issue. it might be cause i built this on an ancient SORT release, but im gonna put some time in it and properly assign it in the trackers.

note to self: dont 📝drink📝and📝code. 

about deep sort, theres alot of implementations of deep sort and yolo but they all rely on torch, keras or tf + reid. how are you thinking of going around that?

im looking at

**Comment by minhtcai** [comment]
Thanks man, I'm out of town this week and will add your part soon :D 
Tbh, I'm reading through these threads as references, hope that I could find an approach:
1. https://github.com/AlexeyAB/darknet/issues/6004
2. https://github.com/Smorodov/Multitarget-tracker
3. https://github.com/GeekAlexis/FastMOT
4. https://github.com/AlexeyAB/darknet/issues/3042
5. https://github.com/nwojke/deep_sort

You can have some coffee though haha
Again thanks man!!

**Comment by akashAD98** [comment]
@minhtcai hey any update on this?

**Comment by sekisek** [comment]
> follow up, after this i looked into the code and there is a clear idswitch issue. it might be cause i built this on an ancient SORT release, but im gonna put some time in it and properly assign it in the trackers.
> 
> note to self: dont 📝drink📝and📝code.
> 
> about deep sort, theres alot of implementations of deep sort and yolo but they all rely on torch, keras or tf + reid. how are you thinking of going around that?
> 
> im looking at

@haviduck 
what u said about deep sort is so on point everything is for other frameworks.

any progress?

**Comment by haviduck** [comment]
> > follow up, after this i looked into the code and there is a clear idswitch issue. it might be cause i built this on an ancient SORT release, but im gonna put some time in it and properly assign it in the trackers.
> > note to self: dont 📝drink📝and📝code.
> > about deep sort, theres alot of implementations of deep sort and yolo but they all rely on torch, keras or tf + reid. how are you thinking of going around that?
> > im looking at
> 
> @haviduck what u said about deep sort is so on point everything is for other frameworks.
> 
> any progress?

yeah. im not done with the repo, but its out there :) working on making it friendlier.

**Comment by sekisek** [comment]
BTW  @haviduck 
did you see:
https://github.com/Smorodov/Multitarget-tracker

?
it looks amazing but its implementation throw python is not completely done...

**Comment by haviduck** [comment]
> BTW @haviduck did you see: https://github.com/Smorodov/Multitarget-tracker
> 
> ? it looks amazing but its implementation throw python is not completely done...

no thats a new one for me. looks interesting gonna have to dive in. 
theres also ByteTrack out which is promising, using yolox.

**Comment by sekisek** [comment]
@haviduck 
ByteTrack ?? it is also with torch, no?
and we don't want that :)
